firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/FirestoreMapEncoder.kt

```diff
@@ -0,0 +1,210 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.ktx.serialization
+
+import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.StructureKind
+import kotlinx.serialization.descriptors.elementNames
+import kotlinx.serialization.encoding.AbstractEncoder
+import kotlinx.serialization.encoding.CompositeEncoder
+import kotlinx.serialization.modules.EmptySerializersModule
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.serializer
+
+/**
+ * The entry point of Firestore Kotlin Serialization Process. It encodes a custom @Serializable
+ * Kotlin object into a nested map of primitives. SDK users don’t need to directly interact with
+ * [FirestoreMapEncoder].
+ *
+ * <p>For a custom @Serializable object, at compile time, a serializer will be generated by the
+ * Kotlin serialization compiler plugin (or a custom serializer can be manually passed in). The
+ * structure information of the custom @Serializable object will be recorded inside of the
+ * serializer’s descriptor (i.e. the name/type of each property to be encoded, the annotation on
+ * each property).
```

Comment:
re: annotations. Once you get access to such annotations, how will firestore specific state be passed into your encoder, i.e. where will documentId come from?